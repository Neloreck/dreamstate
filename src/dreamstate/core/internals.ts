import type { Context } from "react";

import type {
  TAnyContextManagerConstructor,
  TAnyValue,
  TQuerySubscriptionMetadata,
  TSignalSubscriptionMetadata,
} from "@/dreamstate/types";

/**
 * Meta symbols for private internals in context managers.
 */
export const SIGNAL_METADATA_SYMBOL: unique symbol = Symbol("SIGNAL_METADATA");
export const QUERY_METADATA_SYMBOL: unique symbol = Symbol("QUERY_METADATA");
export const SIGNALING_HANDLER_SYMBOL: unique symbol = Symbol("SIGNALING_HANDLER");
export const SCOPE_SYMBOL: unique symbol = Symbol("SCOPE");

/**
 * A weak map registry that stores React context instances bound to specific manager classes.
 *
 * This registry ensures that the library does not retain unnecessary references to manager classes,
 * preventing memory leaks and unintended side effects.
 *
 * This is particularly useful in scenarios such as:
 * - Hot Module Replacement
 * - Module unloading
 * - Scope disposal
 */
export const CONTEXT_REACT_CONTEXTS_REGISTRY: WeakMap<
  TAnyContextManagerConstructor,
  Context<TAnyValue>
> = new WeakMap();

export const SIGNAL_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TSignalSubscriptionMetadata> =
  new WeakMap();

export const QUERY_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TQuerySubscriptionMetadata> =
  new WeakMap();
