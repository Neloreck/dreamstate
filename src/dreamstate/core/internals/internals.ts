import type { Context } from "react";

import type {
  TServiceMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber,
  TDreamstateService,
  TAnyObject
} from "@/dreamstate/types";

/**
 * Global shared constants.
 */

export const EMPTY_ARR: [] = [];

/**
 * Store related constants.
 */

// Currently active services instance types.
export const CONTEXT_SERVICES_ACTIVATED: Set<TDreamstateService> = new Set();

// Currently active services instance types.
export const CONTEXT_SERVICES_REFERENCES: TServiceMap<number> = new WeakMap();

// ReactContext registry, lazy initialized constants.
export const CONTEXT_REACT_CONTEXTS_REGISTRY: TServiceMap<Context<any>> = new WeakMap();

// ContextService instances registry.
export const CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TDreamstateService>> = new WeakMap();

// ContextManagers context references registry.
export const CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject> = new WeakMap();

// ContextManagers observers registry - Providers of related context.
export const CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>> = new WeakMap();

// ContextManagers subscribers registry for data updates.
export const CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>> = new WeakMap();

// Signal handlers registry bound to context instances.
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TServiceMap<TSignalListener<any>> = new WeakMap();

// Signal listeners metadata registry related to services.
export const CONTEXT_SIGNAL_METADATA_REGISTRY: TServiceMap<TSignalSubscriptionMetadata> = new WeakMap();

// Query listeners metadata registry related to services.
export const CONTEXT_QUERY_METADATA_REGISTRY: TServiceMap<TQuerySubscriptionMetadata> = new WeakMap();

// Global signal listeners registry.
export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();
