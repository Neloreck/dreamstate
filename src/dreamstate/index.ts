/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export {
  ILoadable as Loadable,
  TNested as Nested,
  TComputed as Computed,
  TStateSetter as StateSetter,
  TConsumable as Consumable,
  ISignalEvent as SignalEvent,
  ISignal as Signal,
  TSignalType as SignalType,
  TQueryResponse as QueryResponse,
  TOptionalQueryResponse as OptionalQueryResponse,
  TQueryRequest as QueryRequest,
  TQueryType as QueryType
} from "@/dreamstate/types";

export { ContextService } from "@/dreamstate/core/services/ContextService";
export { ContextManager } from "@/dreamstate/core/services/ContextManager";

export { getCurrentContext } from "@/dreamstate/core/registry/getCurrentContext";
export { getCurrent } from "@/dreamstate/core/registry/getCurrent";

export { createProvider } from "@/dreamstate/core/provision/createProvider";
export { Provide } from "@/dreamstate/core/provision/Provide";
export { withProvision } from "@/dreamstate/core/provision/withProvision";

export { OnSignal } from "@/dreamstate/core/signals/OnSignal";
export { useSignals } from "@/dreamstate/core/signals/useSignals";
export { unsubscribeFromSignals } from "@/dreamstate/core/signals/unsubscribeFromSignals";
export { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
export { emitSignal } from "@/dreamstate/core/signals/emitSignal";

export { OnQuery } from "@/dreamstate/core/queries/OnQuery";
export { queryData } from "@/dreamstate/core/queries/queryData";

export { useManager } from "@/dreamstate/core/consumption/useManager";
export { Consume } from "@/dreamstate/core/consumption/Consume";
export { withConsumption } from "@/dreamstate/core/consumption/withConsumption";

export { Bind } from "@/dreamstate/utils/Bind";
export { createLoadable } from "@/dreamstate/utils/createLoadable";
export { createNested } from "@/dreamstate/utils/createNested";
export { createComputed } from "@/dreamstate/utils/createComputed";
export { createSetter } from "@/dreamstate/utils/createSetter";
