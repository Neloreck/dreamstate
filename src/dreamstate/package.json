{
  "name": "dreamstate",
  "version": "4.0.0-experimental.6",
  "description": "Store management library based on context and observers.",
  "sideEffects": false,
  "author": "Syrotenko Igor",
  "homepage": "https://github.com/neloreck/dreamstate#readme",
  "license": "MIT",
  "main": "./index.js",
  "module": "./esm/index.js",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/neloreck/dreamstate.git"
  },
  "bugs": {
    "url": "https://github.com/neloreck/dreamstate/issues"
  },
  "keywords": [
    "action",
    "class-based",
    "declarative",
    "decorator",
    "metadata",
    "react",
    "reducer",
    "redux",
    "state",
    "typescript",
    "utility"
  ],
  "dependencies": {
    "tslib": "^2.1.0",
    "shallow-equal": ">=1.0.0"
  },
  "peerDependencies": {
    "react": ">=16.8.0"
  }
}
