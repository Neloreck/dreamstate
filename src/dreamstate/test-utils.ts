/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export { ScopeContext as InternalScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

export * from "@/dreamstate/test-utils/registry/getCurrent";
export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";

export * from "@/dreamstate/test-utils/registry/mockManagerWithScope";
export * from "@/dreamstate/test-utils/registry/mockScopeProvider";
export * from "@/dreamstate/test-utils/registry/mockScope";

export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";
