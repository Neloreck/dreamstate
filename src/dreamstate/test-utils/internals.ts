/**
 * Meta symbols for private internals in context managers for test-utils.
 */
export const IS_MOCKED: unique symbol = Symbol("IS_MOCKED");
