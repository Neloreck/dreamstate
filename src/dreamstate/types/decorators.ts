import { TConstructor } from "@/dreamstate/types/general";

/**
 * From the TC39 Decorators proposal.
 */
// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ClassElement {
  kind: "field" | "method";
  key: PropertyKey;
  placement: "static" | "prototype" | "own";
  initializer?: Function;
  extras?: ClassElement[];
  finisher?: <T>(clazz: TConstructor<T>) => void | undefined | TConstructor<T>;
  descriptor?: PropertyDescriptor;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ClassDescriptor {
  kind: "class";
  elements: ClassElement[];
  finisher?: <T>(clazz: TConstructor<T>) => void | TConstructor<T>;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface MethodDescriptor extends ClassElement {
  kind: "method";
  descriptor: PropertyDescriptor;
}
