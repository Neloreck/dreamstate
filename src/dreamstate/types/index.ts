export * from "@/dreamstate/types/queries";
export * from "@/dreamstate/types/signals";
export * from "@/dreamstate/types/error";
export * from "@/dreamstate/types/general";
export * from "@/dreamstate/types/decorators";
export * from "@/dreamstate/types/utils";
export * from "@/dreamstate/types/internal";
