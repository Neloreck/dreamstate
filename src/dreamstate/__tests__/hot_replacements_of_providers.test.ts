import { mount } from "enzyme";
import { createElement, ReactElement } from "react";

import { ContextManager, createProvider } from "@/dreamstate";
import { CONTEXT_SERVICES_ACTIVATED } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

describe("Mount order for providers", () => {
  it("Should correctly replace context managers", async () => {
    let provisionsStarted: number = 0;
    let provisionsEnded: number = 0;

    let firstProvided: number = 0;
    let firstRemoved: number = 0;
    let fourthProvided: number = 0;
    let fourthRemoved: number = 0;

    class Base extends ContextManager<TAnyObject> {

      public context: TAnyObject = {};

      protected onProvisionStarted() {
        provisionsStarted += 1;
      }

      protected onProvisionEnded() {
        provisionsEnded += 1;
      }

    }

    class First extends Base {

      protected onProvisionStarted() {
        super.onProvisionStarted();
        firstProvided += 1;
      }

      protected onProvisionEnded() {
        super.onProvisionEnded();
        firstRemoved += 1;
      }

    }

    class Second extends Base {}

    class Third extends Base {}

    class Fourth extends Base {

      protected onProvisionStarted() {
        super.onProvisionStarted();
        fourthProvided += 1;
      }

      protected onProvisionEnded() {
        super.onProvisionEnded();
        fourthRemoved += 1;
      }

    }

    function RootProvisioner({ sources }: { sources: Array<TAnyContextManagerConstructor> }): ReactElement {
      const provider = createProvider(sources);

      return createElement(provider, {}, createElement("div", {}, "nested"));
    }

    let sources = [ First, Second, Third ];
    const node = mount(createElement(RootProvisioner, { sources }));

    expect(node).toMatchSnapshot("Providers tree with HMR");
    expect(provisionsStarted).toBe(3);
    expect(provisionsEnded).toBe(0);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(3);
    expect(CONTEXT_SERVICES_ACTIVATED.has(First)).toBeTruthy();

    node.setProps({ sources });

    expect(provisionsStarted).toBe(6);
    expect(provisionsEnded).toBe(3);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(3);
    expect(CONTEXT_SERVICES_ACTIVATED.has(First)).toBeTruthy();

    sources = [ ...sources ];
    sources[0] = Fourth;

    node.setProps({ sources });

    expect(provisionsStarted).toBe(9);
    expect(provisionsEnded).toBe(6);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(3);
    expect(CONTEXT_SERVICES_ACTIVATED.has(First)).toBeFalsy();
    expect(CONTEXT_SERVICES_ACTIVATED.has(Fourth)).toBeTruthy();

    expect(firstProvided).toBe(2);
    expect(firstRemoved).toBe(2);
    expect(fourthProvided).toBe(1);
    expect(fourthRemoved).toBe(0);

    node.unmount();

    expect(provisionsStarted).toBe(9);
    expect(provisionsEnded).toBe(9);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(0);
  });

  it("Should correctly work with few providers of same services", () => {
    let provisionsStarted: number = 0;
    let provisionsEnded: number = 0;

    class Base extends ContextManager<TAnyObject> {

      public context: TAnyObject = {};

      protected onProvisionStarted() {
        provisionsStarted += 1;
      }

      protected onProvisionEnded() {
        provisionsEnded += 1;
      }

    }

    class First extends Base {}

    class Second extends Base {}

    function RootProvisioner({ sources }: { sources: Array<TAnyContextManagerConstructor> }): ReactElement {
      const provider = createProvider(sources);

      return createElement(
        "div",
        {},
        createElement("div", {}, "nested"),
        createElement(provider, {}, createElement("div", {}, "nested")),
        createElement(provider, {}, createElement("div", {}, "nested")),
        createElement("div", {}, "nested")
      );
    }

    const sources = [ First, Second ];
    const node = mount(createElement(RootProvisioner, { sources }));

    expect(node).toMatchSnapshot("Multiple providers tree");
    expect(provisionsStarted).toBe(2);
    expect(provisionsEnded).toBe(0);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(2);

    node.setProps({ sources });

    expect(provisionsStarted).toBe(4);
    expect(provisionsEnded).toBe(2);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(2);

    node.unmount();

    expect(provisionsStarted).toBe(4);
    expect(provisionsEnded).toBe(4);
    expect(CONTEXT_SERVICES_ACTIVATED.size).toBe(0);
  });
});
