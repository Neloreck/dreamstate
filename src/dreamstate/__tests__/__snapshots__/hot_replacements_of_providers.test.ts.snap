// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount order for providers Should correctly replace context managers: Providers tree with HMR 1`] = `
<RootProvisioner
  sources={
    Array [
      [Function],
      [Function],
      [Function],
    ]
  }
>
  <Memo(DS.Observer)>
    <div>
      nested
    </div>
  </Memo(DS.Observer)>
</RootProvisioner>
`;

exports[`Mount order for providers Should correctly work with few providers of same services: Multiple providers tree 1`] = `
<RootProvisioner
  sources={
    Array [
      [Function],
      [Function],
    ]
  }
>
  <div>
    <div>
      nested
    </div>
    <Memo(DS.Observer)>
      <div>
        nested
      </div>
    </Memo(DS.Observer)>
    <Memo(DS.Observer)>
      <div>
        nested
      </div>
    </Memo(DS.Observer)>
    <div>
      nested
    </div>
  </div>
</RootProvisioner>
`;
